File: rectangular/cell.go
package rectangular

type cell struct {
	row int
	col int
}

func Cell(row, col int) cell {
	return cell{row: row, col: col}
}

func (c cell) Hash() int {
	return CellHash(c)
}

func (c cell) North() cell {
	return cell{c.row - 1, c.col}
}

func (c cell) South() cell {
	return cell{c.row + 1, c.col}
}

func (c cell) East() cell {
	return cell{c.row, c.col + 1}
}

func (c cell) West() cell {
	return cell{c.row, c.col - 1}
}

func CellHash(c cell) int {
	return c.row*1e3 + c.col
}

File: rust/src/rectangular_maze/cell.rs
// A Cell represents a location in the Maze. It can also be thought of as a node
// in a graph.
#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Cell {
    pub x: i16,
    pub y: i16,
}

impl Cell {
    pub fn north(&self) -> Cell {
        Cell {
            x: self.x,
            y: self.y + 1,
        }
    }
    pub fn east(&self) -> Cell {
        Cell {
            x: self.x + 1,
            y: self.y,
        }
    }
    pub fn south(&self) -> Cell {
        Cell {
            x: self.x,
            y: self.y - 1,
        }
    }
    pub fn west(&self) -> Cell {
        Cell {
            x: self.x - 1,
            y: self.y,
        }
    }
}
